### PR TITLE
PieChart: add demo with custom styles (patterns, gradients)

### PR DESCRIPTION
**Pull request by Krijovnick**

### CODE REVIEW

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    if (point.argument === 'Linear Gradient') {
+      pointSettings.color = { fillId: this.linearGradient(color) };
+    }
+    if (point.argument === 'Radial Gradient') {
+      pointSettings.color = { fillId: this.radialGradient(color) };
+    }
+    if (point.argument === 'Pattern with image') {
+      pointSettings.color = { fillId: this.patternImage(color) };
+    }
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts:51]: Let's use switch to make the code more pretty. We can also choose only `fillId` in switch and set `pointSettings.color` only after swithc to reduce a copypaste.

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    return pointSettings;
+  };
+
+  customizeText = (info) => info.argument;
+
+  hexToRgb = (hex: string, opacity = 1) => {
+    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: `result` is too common name for me.
We can also move a regex to a constant with meaningful name to make the code clearer.

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    return `rgba(${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(result[3], 16)}, ${opacity})`;
+  };
+
+  gradient = (type: string, color1: string, color2: string) => registerGradient(type, {
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: Is it necessary to create a new pattern on each `customizePoint` call? Cannot we create it only once in a constructor?

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    colors: [{
+      offset: '20%',
+      color: color1,
+    }, {
+      offset: '90%',
+      color: color2,
+    }],
+  });
+
+  linearGradient = (color: string) => this.gradient('linear', color, this.hexToRgb(color, 0.5));
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: It's better to name function starting a verb.
Maybe `getLinearGradient` or `createLinearGradient`?

File: MVCDemos/DevExtreme.MVC.Demos.csproj [hunk 5]
```diff
@@ -802,6 +802,7 @@
     <Content Include="Content\DemosStyles\Charts\PieWithCustomLabels.css" />
     <Content Include="Content\DemosStyles\Charts\PieWithMultipleSeries.css" />
     <Content Include="Content\DemosStyles\Charts\PieWithSmallValuesGrouped.css" />
+    <Content Include="Content\DemosStyles\Charts\PieWithCustomStyles.css" />
```
Comment (ksercs) [inline MVCDemos/DevExtreme.MVC.Demos.csproj:805]: As far as i remember we should also add a new `png` here.
Please, build the project locally and test that everything is ok

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      container.appendChild(image);
+    },
+  });
+
+  strokePattern = (color: string) => registerPattern({
+    width: this.shapePatternSize,
+    height: this.shapePatternSize,
+    template: (container) => {
+      const halfSize = this.shapePatternSize / 2;
+      const oneHalfSize = this.shapePatternSize * 1.5;
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: @vladaskorohodova  is this name ok?)

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.css [hunk 7]
```diff
@@ -0,0 +1,3 @@
+::ng-deep #pie {
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.css]: Do we really need this ng-deep?

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+      color: color1,
+    }, {
+      offset: '90%',
+      color: color2,
+    }],
+  });
+
+  linearGradient = (color: string) => this.gradient('linear', color, this.hexToRgb(color, 0.5));
+
+  radialGradient = (color: string) => this.gradient('radial', this.hexToRgb(color, 0.5), color);
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: I suggest u to add a `private` to methods which are not used from outside

File: MVCDemos/Content/DemosStyles/Charts/PieWithCustomStyles.css [hunk 9]
```diff
@@ -0,0 +1,3 @@
+#pie {
+    height: 440px;
+}
```
Comment (ksercs) [inline MVCDemos/Content/DemosStyles/Charts/PieWithCustomStyles.css]: an empty line is missed at the end.
Plz fix all such places

File: MVCDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        new { Type = "Stroke Pattern", Value = 1 },
+        new { Type = "Square Pattern", Value = 1 },
+        new { Type = "Linear Gradient", Value = 1 },
+        new { Type = "Radial Gradient", Value = 1 },
+        new { Type = "Pattern with image", Value = 1 }
+    })
+)
+
+<script>
+    var registerGradient = DevExpress.common.charts.registerGradient,
```
Comment (ksercs) [inline MVCDemos/Views/Charts/PieWithCustomStyles.cshtml]: It's a good practice to create a single variable on a single row

File: MVCDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        registerPattern = DevExpress.common.charts.registerPattern;
+
+    var imagePatternSize = 12,
+        shapePatternSize = 6;
+
+    function customizeText(info) {
+		return info.argument;
+	}
```
Comment (ksercs) [inline MVCDemos/Views/Charts/PieWithCustomStyles.cshtml]: miss formatting

File: NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+)
+
+<script>
+    var registerGradient = DevExpress.common.charts.registerGradient,
+        registerPattern = DevExpress.common.charts.registerPattern;
+
+    var imagePatternSize = 12,
+        shapePatternSize = 6;
+
+	function customizeText(info) {
```
Comment (ksercs) [inline NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml]: miss formatting

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts (hunk 3 above)
Comment (Krijovnick) [reply JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: Yes, it is need to pass the color

File: MVCDemos/Views/Charts/PieWithCustomStyles.cshtml (hunk 11 above)
Comment (Krijovnick) [reply MVCDemos/Views/Charts/PieWithCustomStyles.cshtml]: hm, localy I haven't such format

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts (hunk 6 above)
Comment (Krijovnick) [reply JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: fixed in oneAndAHalfSize

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  };
+
+  customizeText = (info) => info.argument;
+
+  private hexToRgb = (hex: string, opacity = 1) => {
+    const hexArray = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
+    return `rgba(${parseInt(hexArray[1], 16)}, ${parseInt(hexArray[2], 16)}, ${parseInt(hexArray[3], 16)}, ${opacity})`;
+  };
+
+  private gradient = (type: string, color1: string, color2: string) => registerGradient(type, {
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: shouldn't it also me named starting with a verb?

**Review by alexanderPolosatov — DISMISSED** [review]

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        break;
+    }
+
+    return { color: { fillId } };
+  };
+
+  customizeText = (info) => info.argument;
+
+  private hexToRgb = (hex: string, opacity = 1) => {
+    const hexArray = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]:
```suggestion
    const hexColorParts = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
```

File: JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    width: this.imagePatternSize,
+    height: this.imagePatternSize,
+    template: (container) => {
+      const rect = this.createRect(this.imagePatternSize, color, '', 0);
+      const image = document.createElementNS('http://www.w3.org/2000/svg', 'image');
+      image.setAttribute('x', '0');
+      image.setAttribute('y', '0');
+      image.setAttribute('width', this.imagePatternSize.toString());
+      image.setAttribute('height', this.imagePatternSize.toString());
+      image.setAttribute('href', '../../../../images/Charts/PieWithCustomStyles/diamond.png');
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/Angular/app/app.component.ts]: As far as i know, designer recommends not using of a `png` in demos. Did u agreed this?

File: JSDemos/Demos/Charts/PieWithCustomStyles/React/App.js [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+function getStrokePattern(color) {
+  return registerPattern({
+    width: shapePatternSize,
+    height: shapePatternSize,
+    template: (container) => {
+      const halfSize = shapePatternSize / 2;
+      const oneAndAHalfSize = shapePatternSize * 1.5;
+      const d = `M ${halfSize} ${-halfSize} L ${-halfSize} ${halfSize} M 0 ${shapePatternSize} L ${shapePatternSize} 0 M ${oneAndAHalfSize} ${halfSize} L ${halfSize} ${oneAndAHalfSize}`;
```
Comment (ksercs) [inline JSDemos/Demos/Charts/PieWithCustomStyles/React/App.js:84]: we split it to a several rows to get rid of this really long row in Angular, can we use this approach in all frameworks?

File: NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 17]
```diff
@@ -0,0 +1,148 @@
+@(
+Html.DevExtreme().PieChart()
```
Comment (ksercs) [inline NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml]: seems it's not a default formatting for our demos

File: NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+	.ID("pie")
+	.CustomizePoint("customizePoint")
+	.Series(s => {
+		s.Add()
```
Comment (ksercs) [inline NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml]: Hm, seems there are too many indents in .NET Core

File: NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml (hunk 12 above)
Comment (ksercs) [reply NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml]: still not resolved

File: NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+	function customizeText(info) {
+        return info.argument;
+    }
+
+    function customizePoint(point) {
+        var color = point.series.getPointsByArg(point.argument)[0].getColor();
+        var fillId;
+        switch (point.argument) {
+            case 'Stroke Pattern':
+            fillId = getStrokePattern(color);
```
Comment (ksercs) [inline NetCoreDemos/Views/Charts/PieWithCustomStyles.cshtml]: missing indent. Please, fix in all frameworks
```suggestion
                fillId = getStrokePattern(color);
```

File: MVCDemos/Views/Charts/PieWithCustomStyles.cshtml [hunk 20]
```diff
@@ -0,0 +1,147 @@
+@(Html.DevExtreme().PieChart()
+    .ID("pie")
+    .CustomizePoint("customizePoint")
+    .Series(s => {
+        s.Add()
+        .ValueField("Value")
```
Comment (ksercs) [inline MVCDemos/Views/Charts/PieWithCustomStyles.cshtml:5]: Please, fix indents everywhere
```suggestion
            .ValueField("Value")
```

File: JSDemos/Demos/Charts/PieWithCustomStyles/React/App.js (hunk 16 above)
Comment (Krijovnick) [reply JSDemos/Demos/Charts/PieWithCustomStyles/React/App.js:84]: In angular it splited by eslint. In other frameworks it does not work

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by ksercs — APPROVED** [review]

**Review by Shuvaevmedia — APPROVED** [review]

**Pull request merged**